### PR TITLE
Test and fix the notebook tester

### LINKED ISSUE #91: Cell test results are not being checked
Currently, no matter what your test does, it will pass:

https://github.com/jpmorganchase/nbcelltests/pull/90/files#diff-af9ad1804dfd75963191f30ccc04129dL112

Addressed in #90

### LINKED ISSUE #94: Tests include all previous cells+tests, but they also all execute in the same kernel
Each test executes all the previous tests, by design (by including the code of all previous tests in the given test method). However, all tests currently _also_ execute in the same kernel (unless you explicitly distribute each test to its own process). With `N` tests, test `n` is therefore run `N+1-n` times in the same kernel (e.g. with 10 tests, the 1st test executes 10 times in the same kernel).

Let me know if this is deliberate, e.g. because starting a kernel per test method is too slow, or for some other reason. ~~(In #90 I have actually made it an option, but the default is to have "fresh kernel per test".)~~

### PR DESCRIPTION
**Pull request by ceball**
# Significant behavior changes

* [x] Fix missing raise in "vendored tests" (from nbval). Fixes #91.
* [x] Ensure cells are not executed multiple times in the same kernel. Fixes #94.

# tests_vendored cleanup

* [x] ~~Support always starting a new kernel per test (with option for previous behavior).~~ Always start a new kernel per test. Future work: optimize by tracking cell execution so that each cell is only executed once in any given context. #105 
* [x] Fix tests_vendored method names and make class methods actually be class methods

# New tests

* [x] ~~Test cumulative cell run state (shared kernel)~~ (we are dropping this current behavior)
* [x] Test cumulative cell run state (fresh kernel)
* [x] Test with new process per test
* [x] Test cell exception
* [x] Test test exception
* [x] Test test failure

# Cleanup internal to this PR

* [x] Clean up test_test.py (e.g. fix the enterprise class names, make it possible to understand what is being done, generate the test notebooks, etc - see TODOs).
* [x] ~~If keeping optional 'multiple executions of a cell in same kernel' behavior, make option easily selectable.~~ we are not keeping this behavior

### CODE REVIEW

File: nbcelltests/lint.py [hunk 1]
```diff
@@ -158,16 +158,17 @@ def run(notebook, executable=None, rules=None):
     if executable:
         exp = ScriptExporter()
         (body, resources) = exp.from_notebook_node(nb)
+        tf = NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding='utf8')
+        tf_name = tf.name
```
Comment (ceball) [inline nbcelltests/lint.py:162]: changes in this file are unrelated to this PR - will move them

**Comment by timkpaine** [comment]
@ceball did you want to upgrade this from draft PR to real PR?

**Comment by ceball** [comment]
@timkpaine before I finish this PR, please tell me your opinion about #94. I.e. is it ok to modify the existing behavior so that each test is now run in a fresh kernel (rather than the current behavior of running all tests in the same kernel*)?

* unless tests are all distributed to independent processes (behavior in this case unchanged: fresh kernel per test)

**Comment by timkpaine** [comment]
@ceball I think its fine, maybe @vidartf has an opinion

**Comment by ceball** [comment]
@timkpaine @vidartf I will proceed with that change, but Vidar and I will review the behavior before the next release so we can revert if necessary. Hope that's ok.

**Comment by ceball** [comment]
I'd prefer a squash merge here (no need to preserve my false accusations about azure pipelines 😂). Actually I prefer squash merges in general as an easy way to have a nice history on master, but I'm not precious about it and understand the drawbacks of squash merge.

**Comment by ceball** [comment]
@vidartf and I have planned out how we'd like to change the behavior of celltests. The planned changes are an optimization on top of this PR, so I have now completely dropped the ability to go back to the existing behavior (i.e. the behavior before this PR). Apart from dropping ability to select old behavior by option, there have been no significant changes to this PR since my most recent comment 3 days ago.

File: nbcelltests/tests/test_test.py [hunk 2]
```diff
@@ -5,9 +5,223 @@
...
+
+# TODO: This test file's manual use of unittest is brittle
+
+
+def _import_from_path(pypath, f):
+    import importlib.util
+    spec = importlib.util.spec_from_file_location(pypath, f)
+    mod = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(mod)
+    return mod
```
Comment (vidartf) [inline nbcelltests/tests/test_test.py:45]: While I appreciate that this is just our testing code, given my previous experiences with importlib, I would strongly prefer this to be tested on the full 3.5 - 3.8 set of Python versions. Not a blocker though, just a heads up that there might be some differences across versions.

File: nbcelltests/tests/test_test.py [hunk 3]
```diff
@@ -5,9 +5,223 @@
...
+    @classmethod
+    def setUpClass(cls):
+        """
+        Generate test file from notebook, then import it, and make the
+        resulting module available as "generated_tests" class attribute.
+        """
+        tf = tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding='utf8')
+        tf_name = tf.name
+        try:
+            cls.generated_tests = _import_from_path("nbcelltests.tests.%s.%s" % (__name__, cls.__name__), run(cls.NBNAME, filename=tf_name))
```
Comment (vidartf) [inline nbcelltests/tests/test_test.py:60]: This is a pretty obtuse line of code. Some comments would be nice :) E.g.:
- Why is the class name included in the import location spec? Is this merely for display purposes, or does it require class names to follow some formula?
- Maybe write a little about what `generated_tests` will be after this is called, and how it should be used.

File: nbcelltests/tests/test_test.py [hunk 4]
```diff
@@ -5,9 +5,223 @@
...
+
+    NBNAME = CUMULATIVE_RUN
+
+    def test_state(self):
+        """
+                    cell    test   state
+        0:           -       -       -
+
+        1: either   x=0      -      x=0   (expected)
+               or    -       -      x>0   (bad; x was previously defined in kernel)
```
Comment (vidartf) [inline nbcelltests/tests/test_test.py:76]: What would trigger the `or` condition of this either/or branching? Can you add a short comment on the top describing how to read the table?

File: nbcelltests/tests_vendored.py [hunk 5]
```diff
@@ -46,10 +46,23 @@
...
-    def teardown_class(self):
+    KERNEL_NAME = "{kernel_name}"
+
+    @classmethod
+    def setUpClass(cls):
+        cls.kernel = RunningKernel(cls.KERNEL_NAME)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.kernel.stop()
```
Comment (vidartf) [inline nbcelltests/tests_vendored.py:58]: What is the class-level kernel used for?

File: setup.py [hunk 6]
```diff
@@ -29,7 +29,8 @@
 dev_requires = requires + [
     'bump2version',
     'mock',
-    'autopep8'
+    'autopep8',
+    'pytest-xdist ; sys_platform != "win32"'
```
Comment (vidartf) [inline setup.py:33]: Is it really needed to use the `--forked` option (i.e. vs. just `-n2`)? If we could just use `-n2`, then the distributed test would also work on Windows. The state of the `--forked` option seems confused as well, as it has been spun out of xdist into pytest-forked, but that is listed as unmaintained.

File: nbcelltests/tests/test_test.py (hunk 2 above)
Comment (ceball) [reply nbcelltests/tests/test_test.py:45]: Ah, there was a TODO comment to exactly this effect. I wonder where it went. I'll restore it!

File: nbcelltests/tests/test_test.py (hunk 4 above)
Comment (ceball) [reply nbcelltests/tests/test_test.py:76]: It's tricky - if you can suggest something better I'll be grateful. 

If `x` is already defined, nothing is done. If `x` is not defined, it's set to 0. This is how I caught the case that the tests were running all the cells multiple times in the same kernel.

File: nbcelltests/tests_vendored.py (hunk 5 above)
Comment (ceball) [reply nbcelltests/tests_vendored.py:58]: We ultimately want one kernel per class (i.e. one kernel per set of tests being run).

The hack we currently have to ensure not executing a cell multiple times in the same kernel is:

```
    # TODO: starting a new kernel per test is expensive, and
    # could be optimized.
    def setUp(self):
        self.kernel = RunningKernel(self.KERNEL_NAME)
```

So, starting a kernel in the `setUpClass` method is the one that will remain - starting a kernel in `setUp` will be removed.

Does that make sense? Or I am confused...?

File: setup.py (hunk 6 above)
Comment (ceball) [reply setup.py:33]: Originally, before I made changes to the code, and had only written tests, I used `--forked` to run the new tests to verify that the new tests passed when there was **one test per process** (because that would guarantee one kernel per test). `--forked` is the way I could see to do that easily. Now I have made changes to the code, `--forked` is not really necessary at all any more and could actually be removed.

After the subsequent optimization is applied, we will be more interested in just distributing the tests in some way like you say with half going somewhere and half going somewhere else. So would you accept your proposed change during that subsequent optimization work instead?

File: nbcelltests/tests/test_test.py (hunk 2 above)
Comment (ceball) [reply nbcelltests/tests/test_test.py:45]: > that this is just our testing code

What kind of attitude is that in a library about testing code, ha ha 😲

Comment (ceball) [reply nbcelltests/tests/test_test.py:45]: It's like how I wanted to turn off the linter while I was working on the linting part.

File: nbcelltests/tests/test_test.py (hunk 3 above)
Comment (ceball) [reply nbcelltests/tests/test_test.py:60]: Agree with your first bullet point (and that the line is obtuse).

For the second bullet point, is the existing method docstring enough?

File: nbcelltests/tests_vendored.py (hunk 5 above)
Comment (ceball) [reply nbcelltests/tests_vendored.py:58]: (Of course, you're right that at the moment the class level kernel is pointless, as it's immediately replaced by each test. But I'd rather keep the class level kernel for now, as it's what should be there if we weren't hacking things with a new kernel per test.)

Comment (vidartf) [reply nbcelltests/tests_vendored.py:58]: Ok, so we keep it in this diff as a forward looking change :)

File: nbcelltests/tests/test_test.py (hunk 4 above)
Comment (vidartf) [reply nbcelltests/tests/test_test.py:76]: > If x is already defined, nothing is done. If x is not defined, it's set to 0. 

How do you read this out of the table though? 😕

Comment (ceball) [reply nbcelltests/tests/test_test.py:76]: I said it was tricky! :) I'll think about it - I'm sure we can do better.

**Comment by ceball** [comment]
@vidartf I believe I've addressed the points you raised; please let me know if I missed anything, or otherwise please explicitly approve (a requirement for merging).

File: nbcelltests/tests/test_test.py [hunk 7]
```diff
@@ -68,21 +68,29 @@ class TestCumulativeRun(_TestCellTests):
...
+        not change x's value. If for some reason each test results in
+        multiple executions of the cells, the results will look like
+        this:
+
+             cell    test   state
+        1:    -       -       -
+        2:    -       -      x>0   (bad; x was previously defined in kernel)
+        3:   x+=1     -      x>1   (bad)
+        4:   x+=1     -      x>2   (bad)
+        5:    -      x+=1    x>3   (bad)
```
Comment (vidartf) [inline nbcelltests/tests/test_test.py:111]: Thanks this is much better :)

File: setup.py (hunk 6 above)
Comment (ceball) [reply setup.py:33]: https://github.com/jpmorganchase/nbcelltests/issues/107
